File: day7/solution.cpp
#include "../getValuesFromFile.h"
#include "../split.h"
#include <vector>
#include <string>
#include <iostream>
using namespace std;

class File;
class Directory;

class File{
    public:
        string name;
        long int size;
        Directory* parent;
};

class Directory{
    public:
        string name;
        vector<File> children;
        vector<Directory*> subDirectories;
        Directory* parent;
};


Directory* findSubDir(Directory* home, string name) {
    for (Directory* d : home->subDirectories) {
        cout << "subdir name: " << d->name << endl;
        if (d->name == name) {
            return d;
        }
    }
}


int main() {
    vector<string> values = getValuesFromFile("input.txt");
    vector<Directory> directories;
    long int doesNotExist = string::npos;
    Directory home;
    home.name = "home";
    Directory* current = &home;

    for (string v : values) {
        if (v.find("dir") != doesNotExist) {
            vector<string> parts = split(v, ' ');
            string name = parts[1];
            Directory d;
            d.name = name;
            directories.push_back(d);
        } 
        else if (v.find("$") != doesNotExist) {
            // this is a command
            int length = directories.size();
            for (int i = 0; i < length; i++) {
                cout << directories[i].name << endl;
                Directory dir = directories[i];
                cout << "Adding " << dir.name << " to " << current->name << endl;
                dir.parent = current;
                current->subDirectories.push_back(&dir);
            }
            directories.clear();

            if (v.find("..") != doesNotExist) {
                current = current->parent;
            } else if (v.find("/") != doesNotExist) {
                current = &home;
            } else if (v.find("cd") != doesNotExist) {
                cout << v << endl;
                vector<string> parts = split(v, ' ');
                string dirname = parts[2];
                cout << dirname << endl;
                if (dirname != ".." && dirname != "/") {
                    cout << current->subDirectories.size() << endl;
                    current = findSubDir(current, dirname);
                }
            }
        }
    }

    return 0;
}

File: day7/solution.py
class Directory:
    def __init__(self):
        self.name = ""
        self.children = []
        self.subdirectories = []
        self.parent = ""
        self.size = 0

    def set_name(self, name):
        self.name = name

    def add_child(self, child):
        self.children.append(child)

    def add_subdirectory(self, subdirectory):
        self.subdirectories.append(subdirectory)

    def set_parent(self, parent):
        self.parent = parent

    def get_name(self):
        return self.name

    def get_children(self):
        return self.children
    
    def get_subdirectories(self):
        return self.subdirectories

    def get_parent(self):
        return self.parent

    def get_size(self):
        return self.size

    def increment_size(self, i):
        self.size += i

    def copy(self):
        ret = Directory()
        ret.set_name(self.name)
        ret.set_parent(self.parent)
        for child in self.get_children():
            ret.add_child(child)

        for subdirectory in self.get_subdirectories():
            ret.add_subdirectory(subdirectory)

        return ret


class File:
    def __init__(self):
        self.name = ""
        self.parent = Directory()
        self.size = 0

    def set_name(self, name):
        self.name = name

    def set_parent(self, parent):
        self.parent = parent

    def set_size(self, size):
        self.size = size

    def get_name(self):
        return name

    def get_parent(self):
        return parent

    def get_size(self):
        return size


def readFromFile():
    ret = []
    with open("input.txt") as f:
        ret = f.readlines()

    return ret

def getDirectory(current, name):
    for subdir in current.get_subdirectories():
        if subdir.get_name() == name:
            return subdir

    return None

def initializeDirectory(lines):
    home = Directory()
    home.set_name("home")
    home.set_parent(None)
    current = home
    for line in lines:
        splitline = line.split()
        if ".." in splitline:
            current = current.get_parent()
        elif "/" in splitline:
            current = home
        elif "cd" in splitline:
            name = splitline[2]
            current = getDirectory(current, name)
        elif "dir" in splitline:
            # it's a directory
            dirname = splitline[1]
            newdir = Directory()
            newdir.set_name(dirname)
            newdir.set_parent(current)
            current.add_subdirectory(newdir)
        elif "$" not in splitline:
            # it's a file
            filename = splitline[1]
            size = splitline[0]
            newfile = File()
            newfile.set_name(filename)
            newfile.set_parent(current)
            newfile.set_size(size)
            temp = current
            while temp:
                temp.increment_size(int(size))
                temp = temp.get_parent()

    return home


def findTopSizes(l, total):
    for item in l:
        if item.get_size() <= 100000:
            total += item.get_size()
        total = findTopSizes(item.get_subdirectories(), total)
    return total

def findSmallestBigEnoughFile(l, size, spaceNeeded):
    for item in l:
        if item.get_size() <= size and item.get_size() >= spaceNeeded:
            size = item.get_size()
        size = findSmallestBigEnoughFile(item.get_subdirectories(), size, spaceNeeded)
    return size


if __name__ == '__main__':
    lines = readFromFile()
    home = initializeDirectory(lines)
    total = 0
    total = findTopSizes([home], total)
    print("Part 1 answer: ", total)

    totalDiskSpace = 70000000
    spaceNeeded = 30000000
    freeSpace = totalDiskSpace - home.get_size()
    spaceNeeded = spaceNeeded - freeSpace
    smallestBigEnoughSize = findSmallestBigEnoughFile([home], totalDiskSpace, spaceNeeded)
    print("Part 2 answer: ", smallestBigEnoughSize)
    